feat(engine): Populating process instance id in jobs

punitdarira opened the pull request:
related to #4205

punitdarira commented:
@yanavasileva,
The AcquirableJobEntity does not contain a processDefinitionId field. Should we add a new field, or is the processDefinitionId set elsewhere? Additionally, could you explain how the job logs will be visible in the Historic Process Definition page and Job Log view once the processDefinitionId is set?

punitdarira commented:
> Hi @punitdarira,
> 
> Thank you for raising this pull request and showing interest in our product.
> 
> > The AcquirableJobEntity does not contain a processDefinitionId field. Should we add a new field, or is the processDefinitionId set elsewhere?
> 
> Why you need the field there? The `processDefinitionId` field should be populated in a similar way to the `JobEntity` here: https://github.com/camunda/camunda-bpm-platform/pull/4334/files#diff-bbebf2d653092eaf76c21343a5d422db745c0f5ab55aeef084435be0e64ab11bR173-R175
> 
> > Additionally, could you explain how the job logs will be visible in the Historic Process Definition page and Job Log view once the processDefinitionId is set?
> 
> In Cockpit, we display all job logs with the specific `processDefinitionId`. Once a job has set this id, the historic job log will inherit it, and out-of-the-box the entities will be populated in the mentioned page and view. No action is required for this comment.
> 
> Lastly, don't forget to cover the made changes with test cases.
> 
> Hope that helps you to continue with the contribution.
> 
> Best, Yana

Hi [Yana](@yanavasileva),
The process definitionId is already being set to the job-
https://github.com/camunda/camunda-bpm-platform/blob/3f6cb3055232c0fa9958e7a351af6ea0eafd7bb2/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobDeclaration.java#L93

The above line gets invoked from 
org.camunda.bpm.engine.impl.batch.AbstractBatchJobHandler.createBatchJob(BatchEntity batch, ByteArrayEntity configuration)
Do we still need to set again in createJobEntities()?

Regarding tests, can you please guide whether we should go for new test file for AbstractBatchJobHandler where we simply mock or do we deploy a bpmn file in a test case and check whether instance and definition ids are set?

yanavasileva commented:
> The process definitionId is already being set to the job

I missed that, however, in debug mode and in general the historic job logs doesn't inherit the process definition id from the batch job entity. At the moment, I can't say why is that. So we need to understand why and fix it eventually.

> Regarding tests, can you please guide

I think we can add a test case(s) to existing classes that test batch execution. Here's an example of single invocation test for set variables we have at the moment: https://github.com/camunda/camunda-bpm-platform/blob/master/engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/SetVariablesBatchTest.java#L673-L695.
examples of interesting batches to cover - migration, modification, set retries

yanavasileva reviewed (changes requested):
As discussed above, the pull request needs more work.

punitdarira commented:
> JobDeclaration.java

Hi [Yana](@yanavasileva),
The property was being set as null as the value was being fetched from ACT_RU_JOBDEF and right now we are not setting the processdefinitionid in this table. I've added the code to get the value from batch configuration. I've added a new test case but the setup of test case is similar to `RestartProcessInstanceAsyncTest.shouldSetExecutionStartTimeInBatchAndHistory` so to avoid running the same setup twice, we can maybe just add a new assert statement to the above mentioned method with a comment.

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/batch/AbstractBatchJobHandler.java:
```diff
@@ -169,6 +171,13 @@ protected void createJobEntities(BatchEntity batch, T configuration, String depl
+
+      if (jobConfiguration.getIds() != null && jobConfiguration.getIds().size() == 1) {
+        job.setProcessInstanceId(jobConfiguration.getIds().get(0));
+        String processDefinitionId = JsonUtil.getString(JsonUtil.asObject(batch.getConfigurationBytes()), PROCESS_DEFINITION_ID, null);
```
❌ This implementation will work only for restart batch operations where we store the process definition . We need solution that should work for all batches.

punitdarira replied:
[Yana](@yanavasileva),
Thank you for your feedback. Let me check and get back to you

punitdarira replied:
[Yana](@yanavasileva ), 
Just out of curiosity, why do we want the property to be set for all the batch types and not just restart? 
For tracking the process instances affected by a batch restart we would only require this property to be set in restart batches right?

yanavasileva replied:
>why do we want the property to be set for all the batch types and not just restart?

The ticket that this issue refers is subtask of https://github.com/camunda/camunda-bpm-platform/issues/4199.
The end goal is to display batch operation details to affected process instances (and definitions where applicable) for **all batch types**.
To give you more context, there's a separate ticket that you also showed interest in working on that's specific for the restart batch operations: https://github.com/camunda/camunda-bpm-platform/issues/4201

yanavasileva commented on engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/RestartProcessInstanceAsyncTest.java:
```diff
@@ -1150,6 +1151,33 @@ public void shouldSetExecutionStartTimeInBatchAndHistory() {
+    HistoricJobLog jobLog = historyService.createHistoricJobLogQuery().jobDefinitionType(Batch.TYPE_PROCESS_INSTANCE_RESTART).list().get(0);
+    System.out.println(jobLog);
+
+    assertEquals(processInstance.getProcessDefinitionId(), jobLog.getProcessDefinitionId());
+    assertEquals(processInstance.getRootProcessInstanceId(), jobLog.getProcessInstanceId());
```
❌ The checks are performed for the historic job log, I would expect to check the jobs as well.

yanavasileva reviewed (changes requested):
The proposed changes cause failures in other batch tests. Also, I would expect the have tests for all batches for which this change make sense.

>we can maybe just add a new assert statement to the above mentioned method with a comment

Yes, you can consider doing that where it make sense add comments (and/or adjust test name) so people are not lost what is tested.

punitdarira commented:
Hi @yanavasileva,
I could think of 3 solutions here.
1. To make sure every applicable batch type adds the process definition id in ACT_RU_JOBDEF. So that by default it gets picked up via https://github.com/camunda/camunda-bpm-platform/blob/3f6cb3055232c0fa9958e7a351af6ea0eafd7bb2/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/JobDeclaration.java#L93
2. Doing the same thing as point 1 but with configuration in ACT_RU_BATCH
3. In AbstractBatchJobHandler.createJobEntities, looking up process definition id from HistoricProcessInstance table using instance id and setting it in the JobEntity.

3rd solution would be the least invasive and clearer. Can you please advice?

yanavasileva commented:
@punitdarira, I am not sure either of the options is great when we think about performance.
For example, at first glance option 3 (fetching the process definition id via the process instance) sounds good, but that implementation would mean for each batch execution job to have an additional query to the database. At the moment, I don't think it will be worth it. I might overlooked that aspect during the creation of the ticket. I will sync with the team and update you and the team with the outcome.

punitdarira commented:
> @punitdarira, I am not sure either of the options is great when we think about performance. For example, at first glance option 3 (fetching the process definition id via the process instance) sounds good, but that implementation would mean for each batch execution job to have an additional query to the database. At the moment, I don't think it will be worth it. I might overlooked that aspect during the creation of the ticket. I will sync with the team then update you and the ticket with the outcome.

sounds good

yanavasileva commented:
@punitdarira, I checked with the team, we want to keep it simple for now and only populate the `processDefinitionId` field for restart batch operations.
Decision: https://github.com/camunda/camunda-bpm-platform/issues/4205#issuecomment-2196882992.

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/batch/AbstractBatchJobHandler.java:
```diff
@@ -169,6 +171,13 @@ protected void createJobEntities(BatchEntity batch, T configuration, String depl
+
+      if (jobConfiguration.getIds() != null && jobConfiguration.getIds().size() == 1) {
+        job.setProcessInstanceId(jobConfiguration.getIds().get(0));
+        String processDefinitionId = JsonUtil.getString(JsonUtil.asObject(batch.getConfigurationBytes()), PROCESS_DEFINITION_ID, null);
+        job.setProcessDefinitionId(processDefinitionId);
```
❌ Considering the last discussions, let's populate the `processDefinitionId` field in `RestartProcessInstancesJobHandler#postProcessJob`
```suggestion
```

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/batch/AbstractBatchJobHandler.java:
```diff
@@ -39,6 +39,8 @@
 import org.camunda.bpm.engine.impl.util.ClockUtil;
 import org.camunda.bpm.engine.impl.util.JsonUtil;
 
+import static org.camunda.bpm.engine.impl.RestartProcessInstancesBatchConfigurationJsonConverter.PROCESS_DEFINITION_ID;
```
❌ Not needed anymore:
```suggestion
```

punitdarira replied:
Hi @yanavasileva .
I've made the changes. I did try adding the test to confirm the instance and definitionId gets set in the job table, but I can't get he hold of restart job entry in this table. When I try to fetch the rows from Job table using - `List<Job> job = managementService.createJobQuery().list();` , I only get the monitor job. I think the job gets deleted soon after execution not sure though. 
I tried this code in `RestartProcessInstanceAsyncTest#shouldSetProcessInstanceAndDefinitionIdInHistoryJobLog`

yanavasileva replied:
I added a comment below to point out where you perform the check for the runtime jobs.

yanavasileva reviewed (changes requested):
Added adjustment for the `processDefinitionId` field.

yanavasileva commented on engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/RestartProcessInstanceAsyncTest.java:
```diff
@@ -1150,6 +1151,32 @@ public void shouldSetExecutionStartTimeInBatchAndHistory() {
+            .executeAsync();
+    helper.executeSeedJob(batch);
+    List<Job> executionJobs = helper.getExecutionJobs(batch);
+
```
✍️ After executing the seed job and before executing the batch execution jobs you can perform the check if `executionJobs.get(0)` has assigned correctly the process instance and definition ids.

punitdarira replied:
Thank you @yanavasileva . I should have thought about that. I've made the change

yanavasileva commented:
@punitdarira, are you willing to add further test coverage that the process instance id is populated for other/all type of batch operations?

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/RestartProcessInstancesJobHandler.java:
```diff
@@ -31,6 +31,9 @@
 import org.camunda.bpm.engine.impl.persistence.entity.JobEntity;
 import org.camunda.bpm.engine.impl.persistence.entity.MessageEntity;
 import org.camunda.bpm.engine.impl.persistence.entity.ProcessDefinitionEntity;
+import org.camunda.bpm.engine.impl.util.JsonUtil;
+
+import static org.camunda.bpm.engine.impl.RestartProcessInstancesBatchConfigurationJsonConverter.PROCESS_DEFINITION_ID;
```
❌ Not needed anymore:
```suggestion
```

punitdarira commented:
> @punitdarira, are you willing to add further test coverage that the process instance id is populated for other/all type of batch operations?

@yanavasileva 
Sure Yana

punitdarira commented:
@yanavasileva ,
Added tests for below batches-
BatchSetVariables				
DecisionSetRemovalTime						  
DeleteHistoricDecisionInstances 
DeleteHistoricProcessInstances	
DeleteProcessInstancesJob		
MessageCorrelationBatchJob		
MigrationBatchJob				
ModificationBatchJob			
ProcessSetRemovalTimeJob					
RestartProcessInstanceBatchJob  
UpdateProcessInstancesSuspendStateJob

However, there were 3 batch types which were operating on batches directly like-
BatchSetRemovalTime
SetExternalTaskRetriesJob
SetJobRetriesJob

For these jobs, the processInstanceId is not being set as expected. I can see the job's id being set instead of the processInstaceId. Just to confirm, the execution jobs for these 3 batches should have the process instacancesId of the transitive instance right?

yanavasileva commented:
Thanks for adding the tests. 💯 
>For these jobs, the processInstanceId is not being set as expected. I can see the job's id being set instead of the processInstaceId. Just to confirm, the execution jobs for these 3 batches should have the process instacancesId of the transitive instance right?

I will check and get back to you. The first one, we won't need a process instance id but I am not sure for the others.

yanavasileva commented on engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/UpdateProcessInstancesSuspendStateAsyncTest.java:
See: `org.camunda.bpm.engine.impl.persistence.entity.JobManager.updateJobSuspensionStateByProcessInstanceId(String, SuspensionState)`

yanavasileva commented on engine/src/test/java/org/camunda/bpm/engine/test/api/runtime/migration/batch/BatchMigrationTest.java:
Note to myself:
Tests failures due to 
```
java.lang.NullPointerException: null
	at org.camunda.bpm.engine.impl.persistence.entity.JobEntity.setExecution(JobEntity.java:259)
	at org.camunda.bpm.engine.impl.persistence.entity.ExecutionEntity.restoreProcessInstance(ExecutionEntity.java:1384)
```

See `org.camunda.bpm.engine.impl.migration.instance.parser.MigratingInstanceParser.fetchJobs(CommandContext, String)`

yanavasileva reviewed (commented):
I am not ready with the review. However, I am sharing my comments to prevent losing them.
Some adjustments needs to be done for suspend and migration batches, I will look at that.

yanavasileva commented on engine/src/main/java/org/camunda/bpm/engine/impl/batch/AbstractBatchJobHandler.java:
```diff
@@ -169,6 +169,11 @@ protected void createJobEntities(BatchEntity batch, T configuration, String depl
 
       JobEntity job = createBatchJob(batch, configurationEntity);
+
+      if (jobConfiguration.getIds() != null && jobConfiguration.getIds().size() == 1) {
```
Something like this will be required to avoid populating the ids for entities that are not process instance:
```suggestion
      if (jobConfiguration.getIds() != null && jobConfiguration.getIds().size() == 1
          && !(this instanceof DecisionSetRemovalTimeJobHandler)
          && !(this instanceof DeleteHistoricDecisionInstancesJobHandler)
          && !(this instanceof SetJobRetriesJobHandler)
          && !(this instanceof SetExternalTaskRetriesJobHandler)
//          && !(this instanceof BatchSetRemovalTimeJobHandler)
          ) {
```

yanavasileva commented:
Just dropping by to say that I didn't have the chance to work on this.
We already started with the other tasks related to this topic so I will provide feedback in the next weeks.
